canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,899 @@
+        pushMetric("TransferRate", transferRate, Aws::CloudWatch::Model::StandardUnit::Kilobits_Second, metricDatum, cusData->pAggregated_dimension, cwRequest);
+        pushMetric("CurrentViewDuration", currentViewDuration, Aws::CloudWatch::Model::StandardUnit::Milliseconds, metricDatum, cusData->pAggregated_dimension, cwRequest);
+        pushMetric("ContentStoreAvailableSize", availableStoreSize, Aws::CloudWatch::Model::StandardUnit::Kilobytes, metricDatum, cusData->pAggregated_dimension, cwRequest);
+    }
+
+    // Capture error rate metrics every 60 seconds
+    double duration = duration_cast<seconds>(system_clock::now().time_since_epoch()).count() - cusData->timeCounter;
+    if(duration > 60)
+    {
+        cusData->timeCounter = duration_cast<seconds>(system_clock::now().time_since_epoch()).count();
+
+        double newPutFrameErrors = (double)stream_metrics_raw->putFrameErrors - cusData->totalPutFrameErrorCount;
+        cusData->totalPutFrameErrorCount = stream_metrics_raw->putFrameErrors;
+        double putFrameErrorRate = newPutFrameErrors / (double)duration;
+        pushMetric("PutFrameErrorRate", putFrameErrorRate, Aws::CloudWatch::Model::StandardUnit::Count_Second, metricDatum, cusData->pDimension_per_stream, cwRequest);
+        LOG_DEBUG("PutFrame Error Rate: " << putFrameErrorRate);
+
+        double newErrorAcks = (double)stream_metrics_raw->errorAcks - cusData->totalErrorAckCount;
+        cusData->totalErrorAckCount = stream_metrics_raw->errorAcks;
+        double errorAckRate = newErrorAcks / (double)duration;
+        pushMetric("ErrorAckRate", errorAckRate, Aws::CloudWatch::Model::StandardUnit::Count_Second, metricDatum, cusData->pDimension_per_stream, cwRequest);
+        LOG_DEBUG("Error Ack Rate: " << errorAckRate);
+
+        double totalNumberOfErrors = cusData->totalPutFrameErrorCount + cusData->totalErrorAckCount;
+        pushMetric("TotalNumberOfErrors", totalNumberOfErrors, Aws::CloudWatch::Model::StandardUnit::Count, metricDatum, cusData->pDimension_per_stream, cwRequest);
+        LOG_DEBUG("Total Number of Errors: " << totalNumberOfErrors);
+
+        if (cusData->pCanaryConfig->useAggMetrics)
+        {
+            pushMetric("PutFrameErrorRate", putFrameErrorRate, Aws::CloudWatch::Model::StandardUnit::Count_Second, metricDatum, cusData->pAggregated_dimension, cwRequest);
+            pushMetric("ErrorAckRate", errorAckRate, Aws::CloudWatch::Model::StandardUnit::Count_Second, metricDatum, cusData->pAggregated_dimension, cwRequest);
+            pushMetric("TotalNumberOfErrors", totalNumberOfErrors, Aws::CloudWatch::Model::StandardUnit::Count, metricDatum, cusData->pAggregated_dimension, cwRequest);
+        }
+
+        cusData->pCanaryLogs->canaryStreamSendLogs(cusData->pCloudwatchLogsObject);
+    }
+
+    // Send metrics to CW
+    auto outcome = cusData->pCWclient->PutMetricData(cwRequest);
```

Comment:
Replace with Async call.

Reply:
Also, what metric is being pushed here? Shouldnt this be done for every stream metric and other metrics?

Reply:
All the metrics are added to the cwRequest to be sent all together in the cwRequest, and it is done in four places in the code: upon persisted ack, upon received ack, upon keyframe, and for startup latency.

Reply:
(In the case above, it is for all the metrics upon keyframe)

Reply:
And for AmazonCloudWatchAsync calls I am getting: 
`error: no matching function for call to ‘Aws::CloudWatch::CloudWatchClient::PutMetricDataAsync(Aws::CloudWatch::Model::PutMetricDataRequest&)`
Is there a different namespace I should be using that includes this? Everything I'm seeing online on this is for Java.